File: 06__full-stack/brainrot-studio/backend/app/models/qa.py
from datetime import datetime, timezone
from enum import Enum
from uuid import UUID, uuid4

from sqlmodel import Field, SQLModel


class QACategory(str, Enum):
    MEDIA = "media"
    TIMELINE = "timeline"
    AUDIO = "audio"
    CAPTIONS = "captions"
    ASSETS = "assets"
    STORY = "story"
    VISUAL = "visual"


class QASeverity(str, Enum):
    INFO = "info"
    WARNING = "warning"
    ERROR = "error"
    CRITICAL = "critical"


class RepairType(str, Enum):
    TIMELINE = "timeline"
    ASSET = "asset"
    DIALOGUE = "dialogue"
    VOICE = "voice"
    CAPTION = "caption"
    AUDIO = "audio"
    SCENE = "scene"
    RENDER = "render"


class QAReport(SQLModel, table=True):
    __tablename__ = "qa_reports"

    id: UUID = Field(default_factory=uuid4, primary_key=True)
    generation_job_id: UUID = Field(foreign_key="generation_jobs.id", index=True)
    attempt: int = Field(default=1)
    score: float = Field(default=100.0)
    passed: bool = Field(default=True, index=True)
    checks_run: int = Field(default=0)
    created_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc),
        nullable=False,
    )


class QAIssueRecord(SQLModel, table=True):
    __tablename__ = "qa_issues"

    id: UUID = Field(default_factory=uuid4, primary_key=True)
    qa_report_id: UUID = Field(foreign_key="qa_reports.id", index=True)
    category: QACategory = Field(index=True)
    severity: QASeverity = Field(index=True)
    code: str = Field(max_length=100, index=True)
    message: str
    scene_id: UUID | None = Field(default=None, foreign_key="scenes.id", nullable=True)
    repairable: bool = Field(default=True)


class RepairAttempt(SQLModel, table=True):
    __tablename__ = "repair_attempts"

    id: UUID = Field(default_factory=uuid4, primary_key=True)
    generation_job_id: UUID = Field(foreign_key="generation_jobs.id", index=True)
    qa_report_id: UUID | None = Field(default=None, foreign_key="qa_reports.id", nullable=True)
    repair_type: RepairType = Field(index=True)
    scene_id: UUID | None = Field(default=None, foreign_key="scenes.id", nullable=True)
    status: str = Field(default="completed", max_length=50)
    reason: str
    created_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

File: 06__full-stack/brainrot-studio/frontend/src/lib/api/qa.ts
import { api } from './client';

export interface QAIssue {
	id: string;
	category: string;
	severity: 'info' | 'warning' | 'error' | 'critical';
	code: string;
	message: string;
	scene_id: string | null;
	repairable: boolean;
}

export interface QAReportResponse {
	id?: string;
	job_id: string;
	attempt: number;
	score: number;
	passed: boolean;
	checks_run: number;
	created_at?: string;
	issues: QAIssue[];
}

export interface QAHistoryResponse {
	job_id: string;
	reports: Array<{
		id: string;
		attempt: number;
		score: number;
		passed: boolean;
		checks_run: number;
		created_at: string;
	}>;
	repair_attempts: Array<{
		id: string;
		repair_type: string;
		scene_id: string | null;
		status: string;
		reason: string;
		created_at: string;
	}>;
}

export function getLatestQAReport(projectId: string, videoId: string, jobId: string) {
	return api<QAReportResponse>(
		`/projects/${projectId}/videos/${videoId}/generation-jobs/${jobId}/qa`
	);
}

export function getQAHistory(projectId: string, videoId: string, jobId: string) {
	return api<QAHistoryResponse>(
		`/projects/${projectId}/videos/${videoId}/generation-jobs/${jobId}/qa/history`
	);
}

export function repairScene(
	projectId: string,
	videoId: string,
	sceneId: string,
	instruction?: string,
	repairType: string = 'SCENE'
) {
	return api<{ status: string; repaired_scene_id: string }>(
		`/projects/${projectId}/videos/${videoId}/scenes/${sceneId}/repair`,
		{
			method: 'POST',
			body: JSON.stringify({ instruction, repair_type: repairType })
		}
	);
}

export function repairJob(jobId: string, sceneId?: string, repairType: string = 'dialogue', reason?: string) {
	return api<{ status: string; job_id: string }>(`/generations/${jobId}/repair`, {
		method: 'POST',
		body: JSON.stringify({ scene_id: sceneId, repair_type: repairType, reason })
	});
}


